### ISSUE TITLE
[sdk/nodejs] Regression: v3.108.0 of `@pulumi/pulumi` breaks Node.js provider SDK TypeScript compilation

### ISSUE THREAD

**Issue by justinvp** [pulumi/pulumi#15556]
Getting the following tsc errors compiling `pulumi-random`'s Node SDK with v3.108.0 of `@pulumi/pulumi`:

```
$ /Users/bryce/src/pulumi-random/sdk/nodejs/node_modules/.bin/tsc
node_modules/@types/glob/index.d.ts:29:42 - error TS2694: Namespace '"/Users/bryce/src/pulumi-random/sdk/nodejs/node_modules/minimatch/dist/cjs/index"' has no exported member 'IOptions'.

29     interface IOptions extends minimatch.IOptions {
                                            ~~~~~~~~

node_modules/@types/glob/index.d.ts:75:30 - error TS2724: '"/Users/bryce/src/pulumi-random/sdk/nodejs/node_modules/minimatch/dist/cjs/index"' has no exported member named 'IMinimatch'. Did you mean 'Minimatch'?

75         minimatch: minimatch.IMinimatch;
```

This appears to be due to https://github.com/pulumi/pulumi/pull/15426.

If we had included https://github.com/pulumi/pulumi/pull/15544 in the release, then this error wouldn't happen.

---

Note that `pulumi-random` is using a [more recent version of TypeScript](https://github.com/pulumi/pulumi-random/blob/090f262dd7015e0ed506675b5440579b4fdf8f50/sdk/nodejs/package.json#L20):

```json
        "typescript": "^4.3.5"
``` 

If I just run `pulumi new typescript` and then run `node_modules/.bin/tsc`, I get a different set of errors, which seem to be related to the fact that by default we include a very old version of TypeScript with `@pulumi/pulumi` (3.8.3):

https://github.com/pulumi/pulumi/blob/940372b66f0ca20787fa90c5ea3cff3359bb92e4/sdk/nodejs/package.json#L42

```
node_modules/@types/node/ts4.8/crypto.d.ts:3534:17 - error TS1110: Type expected.

3534     type UUID = `${string}-${string}-${string}-${string}-${string}`;
                     ~~~

node_modules/@types/node/ts4.8/events.d.ts:101:28 - error TS1005: ',' expected.

101     type AnyRest = [...args: any[]];
                               ~

node_modules/@types/node/ts4.8/test.d.ts:680:34 - error TS1005: '?' expected.

680             : F extends abstract new(...args: any) => infer T ? T
                                     ~~~

node_modules/@types/node/ts4.8/test.d.ts:680:63 - error TS1005: ':' expected.

680             : F extends abstract new(...args: any) => infer T ? T
                                                                  ~

node_modules/@types/node/ts4.8/test.d.ts:681:13 - error TS1005: ',' expected.

681             : unknown,
                ~

node_modules/@types/node/ts4.8/test.d.ts:683:34 - error TS1005: '?' expected.

683             : F extends abstract new(...args: infer Y) => any ? Y
                                     ~~~

node_modules/@types/node/ts4.8/test.d.ts:683:63 - error TS1005: ':' expected.

683             : F extends abstract new(...args: infer Y) => any ? Y
                                                                  ~

node_modules/@types/node/ts4.8/test.d.ts:684:13 - error TS1005: ',' expected.

684             : unknown[],
                ~

node_modules/@types/node/ts4.8/test.d.ts:684:22 - error TS1005: ',' expected.

684             : unknown[],
                         ~

node_modules/@types/node/ts4.8/test.d.ts:685:5 - error TS1109: Expression expected.

685     > {
        ~

node_modules/@types/node/ts4.8/test.d.ts:689:24 - error TS1005: ',' expected.

689         arguments: Args;
                           ~

node_modules/@types/node/ts4.8/test.d.ts:693:35 - error TS1005: ',' expected.

693         error: unknown | undefined;
                                      ~

node_modules/@types/node/ts4.8/test.d.ts:699:39 - error TS1005: ',' expected.

699         result: ReturnType | undefined;
                                          ~

node_modules/@types/node/ts4.8/test.d.ts:703:21 - error TS1005: ',' expected.

703         stack: Error;
                        ~

node_modules/@types/node/ts4.8/test.d.ts:708:19 - error TS1005: ',' expected.

708         target: F extends abstract new(...args: any) => any ? F : undefined;
                      ~~~~~~~

node_modules/@types/node/ts4.8/test.d.ts:708:27 - error TS1005: ':' expected.

708         target: F extends abstract new(...args: any) => any ? F : undefined;
                              ~~~~~~~~

node_modules/@types/node/ts4.8/test.d.ts:708:36 - error TS1005: ',' expected.

708         target: F extends abstract new(...args: any) => any ? F : undefined;
                                       ~~~

node_modules/@types/node/ts4.8/test.d.ts:708:54 - error TS1005: '{' expected.

708         target: F extends abstract new(...args: any) => any ? F : undefined;
                                                         ~~

node_modules/@types/node/ts4.8/test.d.ts:708:63 - error TS1005: ',' expected.

708         target: F extends abstract new(...args: any) => any ? F : undefined;
                                                                  ~

node_modules/@types/node/ts4.8/test.d.ts:708:76 - error TS1005: ',' expected.

708         target: F extends abstract new(...args: any) => any ? F : undefined;
                                                                               ~

node_modules/@types/node/ts4.8/test.d.ts:712:22 - error TS1005: ',' expected.

712         this: unknown;
                         ~

node_modules/@types/node/ts4.8/util.d.ts:1547:42 - error TS1005: ',' expected.

1547         entries(): IterableIterator<[name: string, value: string]>;
                                              ~

node_modules/@types/node/ts4.8/util.d.ts:1547:44 - error TS1068: Unexpected token. A constructor, method, accessor, or property was expected.

1547         entries(): IterableIterator<[name: string, value: string]>;
                                                ~~~~~~

node_modules/@types/node/ts4.8/util.d.ts:1547:57 - error TS1005: ';' expected.

1547         entries(): IterableIterator<[name: string, value: string]>;
                                                             ~

node_modules/@types/node/ts4.8/util.d.ts:1547:65 - error TS1005: ';' expected.

1547         entries(): IterableIterator<[name: string, value: string]>;
                                                                     ~

node_modules/@types/node/ts4.8/util.d.ts:1547:66 - error TS1109: Expression expected.

1547         entries(): IterableIterator<[name: string, value: string]>;
                                                                      ~

node_modules/@types/node/ts4.8/util.d.ts:1547:67 - error TS1109: Expression expected.

1547         entries(): IterableIterator<[name: string, value: string]>;
                                                                       ~

node_modules/@types/node/ts4.8/util.d.ts:1552:17 - error TS1005: ',' expected.

1552         get(name: string): string | null;
                     ~

node_modules/@types/node/ts4.8/util.d.ts:1552:26 - error TS1005: ';' expected.

1552         get(name: string): string | null;
                              ~

node_modules/@types/node/ts4.8/util.d.ts:1556:17 - error TS1005: ',' expected.

1556         has(name: string): boolean;
                     ~

node_modules/@types/node/ts4.8/util.d.ts:1556:26 - error TS1005: ';' expected.

1556         has(name: string): boolean;
                              ~

node_modules/@types/node/ts4.8/util.d.ts:1560:15 - error TS1005: ';' expected.

1560         keys(): IterableIterator<string>;
                   ~

node_modules/@types/node/ts4.8/util.d.ts:1560:41 - error TS1005: '(' expected.

1560         keys(): IterableIterator<string>;
                                             ~

node_modules/@types/node/ts4.8/util.d.ts:1566:17 - error TS1005: ',' expected.

1566         set(name: string, value: string): void;
                     ~

node_modules/@types/node/ts4.8/util.d.ts:1566:32 - error TS1005: ',' expected.

1566         set(name: string, value: string): void;
                                    ~

node_modules/@types/node/ts4.8/util.d.ts:1566:41 - error TS1005: ';' expected.

1566         set(name: string, value: string): void;
                                             ~

node_modules/@types/node/ts4.8/util.d.ts:1566:47 - error TS1109: Expression expected.

1566         set(name: string, value: string): void;
                                                   ~

node_modules/@types/node/ts4.8/util.d.ts:1570:17 - error TS1005: ';' expected.

1570         values(): IterableIterator<string>;
                     ~

node_modules/@types/node/ts4.8/util.d.ts:1570:43 - error TS1005: '(' expected.

1570         values(): IterableIterator<string>;
                                               ~

node_modules/@types/node/ts4.8/util.d.ts:1574:26 - error TS1005: ';' expected.

1574         [Symbol.iterator]: typeof MIMEParams.prototype.entries;
                              ~

node_modules/@types/node/ts4.8/util.d.ts:1576:1 - error TS1128: Declaration or statement expected.

1576 }
     ~

node_modules/minimatch/dist/cjs/ast.d.ts:17:44 - error TS1005: ',' expected.

17     toRegExpSource(allowDot?: boolean): [re: string, body: string, hasMagic: boolean, uflag: boolean];
                                              ~

node_modules/minimatch/dist/cjs/ast.d.ts:17:58 - error TS1005: ',' expected.

17     toRegExpSource(allowDot?: boolean): [re: string, body: string, hasMagic: boolean, uflag: boolean];
                                                            ~

node_modules/minimatch/dist/cjs/ast.d.ts:17:76 - error TS1005: ',' expected.

17     toRegExpSource(allowDot?: boolean): [re: string, body: string, hasMagic: boolean, uflag: boolean];
                                                                              ~

node_modules/minimatch/dist/cjs/ast.d.ts:17:92 - error TS1005: ',' expected.

17     toRegExpSource(allowDot?: boolean): [re: string, body: string, hasMagic: boolean, uflag: boolean];


Found 45 errors.
```

Notes:

1. The `@types/node` errors are coming from the [template](https://github.com/pulumi/templates/blob/934f82f531f8652524c51abd6770530406ff1b83/typescript/package.json#L5)'s use of `"@types/node": "^18"` devDependency. It looks like at some point that package moved to using `ts4.8` features, based on the directory name.

2. The `minimatch` errors would be new with https://github.com/pulumi/pulumi/pull/15426

We should probably open an issue in the templates repo on (1). We may be able to change how the template is defined to avoid using `@types/node` `^18` or using a different tsconfig.json based off of https://github.com/tsconfig/bases, to avoid those errors. Or we just start including a more recent TypeScript version in all of our templates, which we're currently looking into already with https://github.com/pulumi/pulumi/issues/12117.

I'm not sure what we'd do about (2). It is generally unusual for someone to be running `tsc` directly in this way with our templates. Generally, folks just use the built-in `ts-node` capability, which doesn't hit this (because the compilation happens on-demand). And if someone was using `tsc` to manually compile, they'd likely be using a more recent version of typescript (like what `pulumi-random` and other providers do) that doesn't have these problems.

**Comment by justinvp** [comment]
Also hitting this in a PR now:

https://github.com/pulumi/pulumi/actions/runs/8107787381/job/22160093702?pr=15555#step:38:181

```
=== FAIL: . TestLanguage/l2-resource-simple (6.30s)
$ npm install
npm WARN deprecated @opentelemetry/api-metrics@0.32.0: Please use @opentelemetry/api >= 1.3.0

added 301 packages, and audited 302 packages in 3s

42 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
$ npm run build

> @pulumi/simple@$***VERSION*** build
> tsc

##[debug]Dropping file value '/home/runner/work/pulumi/pulumi/node_modules/@types/glob/index.d.ts'. Path does not exist
Error: node_modules/@types/glob/index.d.ts(29,42): error TS2694: Namespace '"/tmp/TestLanguage2516998591/002/sdks/simple-2.0.0/node_modules/minimatch/dist/cjs/index"' has no exported member 'IOptions'.
##[debug]Dropping file value '/home/runner/work/pulumi/pulumi/node_modules/@types/glob/index.d.ts'. Path does not exist
Error: node_modules/@types/glob/index.d.ts(75,30): error TS2724: '"/tmp/TestLanguage2516998591/002/sdks/simple-2.0.0/node_modules/minimatch/dist/cjs/index"' has no exported member named 'IMinimatch'. Did you mean 'Minimatch'?
    language_test.go:186: 
        	Error Trace:	/home/runner/work/pulumi/pulumi/sdk/nodejs/cmd/pulumi-language-nodejs/language_test.go:186
        	Error:      	Received unexpected error:
        	            	rpc error: code = Unknown desc = sdk packing for simple: npm run build: exit status 2
        	Test:       	TestLanguage/l2-resource-simple
```

**Comment by justinvp** [comment]
We should just go ahead and get https://github.com/pulumi/pulumi/pull/15544 merged and released in v3.108.1. I think we can follow-up on what we want to do about (1) and (2) after (likely opening separate issues for those).
